tls/understand-tls/understand-tls.tex

```diff
@@ -0,0 +1,320 @@
+cipher (\lstinline!AEAD!).
+\end{itemize}
+
+\section{Records}
+
+Data is transferred between clients and servers using \emph{records}. Think of 
+a record as being like an envelope containing data. The envelope has some basic 
+information about its contents written on it, such as:
+\begin{itemize}
+\item The amount of data that is being transferred
+\item The type of data that is being transferred
+\end{itemize}
+
+The record will also apply any cryptographic operations to the data that may be 
+appropriate dependant on the current state of the connection. During initial 
+connection various parameters need to be agreed between the two parties to 
+determine exactly what cryptographic operations will be applied. One of these 
+parameters is the ciphersuite. The ciphersuite that is currently in use will 
+define the encryption that is to be applied to the data within the record.
+
+As well as the encrypted data the record will also contain a Message 
+Authentication Code (MAC). The MAC utilises the integrity algorithm and a 
+secret key, shared between the two parties, to calculate a code that is unique
+to the data being sent. Any attempt by an attacker to modify the data will mean
+that the MAC code will fail to verify when it is checked by the remote party 
+and the connection will be aborted.
+
+Optionally records can also compress the data that they transmit. This is done 
+prior to encryption. This is usually not done in practice due to security 
+concerns associated with this capability.
+
+\section{The Handshake}
+
+The initial exchange of messages during which cryptographic parameters are 
+exchanged is known as the \emph{handshake}. While making the initial connection
+no crypotgraphic parameters will have yet been agreed. Therefore there is no
+encryption and no MAC on individual records.  Record data is sent in
+\emph{plaintext}. For this reason the protocol has been designed such that no
+application data is sent until the connection has been established and
+cryptographic parameters have been agreed. A MAC of the entire set of handshake 
+messages is calculated and verified at the last step of the handshake process. 
+In this way, even though inidividual records do not have integrity protection, 
+the handshake as a whole does.
+
+Handshake messages are transmitted between the client and server in rec-ords. A 
+single record may contain multiple handshake messages, or a single handshake 
+message may be spread across multiple records. A handshake is always initiated 
+by a client sending a ``ClientHello'' message to a server. A typical handshake 
+is shown in figure \ref{fig:typical-hand}.
+
+\begin{figure}[t]
+\fbox{\includegraphics[width=0.9\textwidth]{tls/understand-tls/typicalhand.pdf}}
+\caption{A typical SSL/TLS handshake.}
+\label{fig:typical-hand}
+\end{figure}
+
+A typical ClientHello contains:
+\begin{itemize}
+\item The highest protocol version supported by the client or, for TLS 1.3, a
```

Comment:
Would be helpful if these bullet points were numbered. Then the graphic could be labeled and described better.

Reply:
Disagree. This bullet point list describes just the initial ClientHello. As such they all apply to the very first arrow on the diagram - so I don't think it would be helpful to add additional labels to the diagram.

Reply:
Ah, gotcha